### DISCUSSION TITLE
Exponential young modulus increase and hardening

### DISCUSSION THREAD

**Discussion by kschnelll** [thelfer/tfel#347] [Q&A]
Dear Thomas, 

my colleagues and I are currently working on a material model, where 
1. the elastic modulus in z-direction exponentially increases with the total strain in z-direction
2. hardening in the plastic regime exponentially increases with the plastic strain in z-direction
The first part (exponential increase of the Yound Modulus) is working by adding the exponential equation directly into the `@ComputeStiffnessTensor` keyword, which works but leads to an oscillating behavior in the plastic regime, because the elastic modulus is only updated every substep insted of every iteration
Is there a way to update the elastic modulus every iteration? We are using the `Implicit` scheme... 
 
With the second part, we are also facing some issues we were not able to solve so far.
So from my understanding, we would have to change the plasticity part in the `@Integrator` block. 
The first line would look something like this: `fp = (seq - s0 * exp(a*(p+theta*dp)))/D(0,0);`
and the second one like this: `dfp_ddp = theta * s0 * exp(a*(p+theta*dp));`
However, if we try to calculate this model, the solver says "solution failed to converge" in 0 iterations as soon as our load reaches the plastic limit. 
We are using MFront in combination with FEniCs. 

We would be very thankful for every help! 

Best regards, 
Klaus

**Comment by thelfer** [comment]
Hi @kschnelll ,
First of all, thanks to your interest for MFront.
I can't figure out how you managed 1). Do you compute the correction on the FEniCS side ? However, it shall change a lot of things in the computation of the consistent tangent operator.
For 2), you are using the equivalent plastic strain, which is not the axial plastic strain. Don't know if it is related, but there is at least three errors in your derivative: it shall be `dfp_ddp = - a * theta * s0 * exp(a*(p+theta*dp))/D(0,0);`

**Comment by kschnelll** [comment]
Hey Thomas, 
thanks a lot for your quick answer: 
For the first point we just used: 
`@ComputeStiffnessTensor<UnAltered> {
  6000, 3000, "e_z_0 * exp(b * abs(eto_zz))",
  0.2, 0.02, 0.01,
  1667, 125, 125
};`
on the FEniCs side there's no computation of any material properties. 

For 2) we also tried using `@StateVariable real epl_zz;
epl_zz.setEntryName("PlasticStrainZZ");` where epl_zz is defined as `eto[2]-eel[2]`, but this didn't work either, so we thought maybe it works if we go one step back, using `p`, a predefined state variable. 
So obviously, my derivate was wrong, thanks for correcting me! Unfortunately, it didn't change anything...

**Reply by thelfer** [reply]
Can you share the code ?

**Reply by kschnelll** [reply]
Sure, you can find it attached here. It's a bit messy, but I think you will figure it out. 
Looking forward to hear your thoughts

[20221122_GitHub_helfer.txt](https://github.com/thelfer/tfel/files/10067403/20221122_GitHub_helfer.txt)

**Reply by thelfer** [reply]
Ok, there are a few things that I don't understand.

1. Why is `helper` oriented in the opposite direction of the normal in tension (defined by the derivative of abs(s[2]) with respect to the stress) ?  I am pretty sure this leads to an unstable material in tension... But as in the `@InitLocalVariables` code block, you seems to expect  compacting only in compression. Thus there is at least an inconsistency
2. You have two plastic surfaces. Elastic prediction is no enougth in this case: you shall use a status method to be sure that you converge to the right value. You shall also have **two** plastic multipliers
3. There is a lot of mistakes related to the normal and its derivative in the compaction case

Please find attached a simplified example of the compaction part that seems to works (MFront implementation and MTest test in uniaxial compression)

[test.zip](https://github.com/thelfer/tfel/files/10069735/test.zip)

**Reply by kschnelll** [reply]
Thanks for your quick answer and your code snippet. However if I try to run it, I can't use `p_z` instead of `p`. Is there any documentation on which pre-defined variables I can access? 
1. That was something I changed, trying to find the mistake... Obviously, `helper` needs to be positive as you mentioned
2. Can you explicate this point a bit more? I don't really understand what you mean by that...

I think I figured out a way that works for us in the compaction case. The results look fine as long as I'm using a constant elastic modulus in z direction. As soon as I change it to an exponential function it also works for loading of the material, but during unloading, I get an oscillating behavior (see the attached image). From my understanding this has to do with the stiffness tensor being computed only once per timestep and not once per iteration. Do you agree on this or have another explanation? And if so, is there a way to update the stiffness tensor inside the `@Integrator` block? 
![displacement_z](https://user-images.githubusercontent.com/118770163/203816730-6b8b7878-1a35-4dee-a434-dcac047d9a67.png)

**Reply by thelfer** [reply]
> Thanks for your quick answer and your code snippet. However if I try to run it, I can't use `p_z` instead of `p`. 

Can't say without looking a the code :)

> Is there any documentation on which pre-defined variables I can access?

In french, unfortunately. This is more or less described in most tutorials available here: https://thelfer.github.io/tfel/web/gallery.html

> 1. That was something I changed, trying to find the mistake... Obviously, `helper` needs to be positive as you mentioned
> 2. Can you explicate this point a bit more? I don't really understand what you mean by that...

You may want to read this tutorial: https://thelfer.github.io/tfel/web/iwan.html

> I think I figured out a way that works for us in the compaction case. The results look fine as long as I'm using a constant elastic modulus in z direction. As soon as I change it to an exponential function it also works for loading of the material, but during unloading, I get an oscillating behavior (see the attached image). From my understanding this has to do with the stiffness tensor being computed only once per timestep and not once per iteration. Do you agree on this or have another explanation? And if so, is there a way to update the stiffness tensor inside the `@Integrator` block?

Probably, yes. If you can't live with that (I would not), then we must make things a bit differently (the implementation of the consistent tangent operator becomes a more complex)

**Reply by thelfer** [reply]
> Thanks for your quick answer and your code snippet. However if I try to run it, I can't use `p_z` instead of `p`. 

Can't say without looking a the code :)

> Is there any documentation on which pre-defined variables I can access?

In french, unfortunately. This is more or less described in most tutorials available here: https://thelfer.github.io/tfel/web/gallery.html

> 1. That was something I changed, trying to find the mistake... Obviously, `helper` needs to be positive as you mentioned
> 2. Can you explicate this point a bit more? I don't really understand what you mean by that...

You may want to read this tutorial: https://thelfer.github.io/tfel/web/iwan.html

> I think I figured out a way that works for us in the compaction case. The results look fine as long as I'm using a constant elastic modulus in z direction. As soon as I change it to an exponential function it also works for loading of the material, but during unloading, I get an oscillating behavior (see the attached image). From my understanding this has to do with the stiffness tensor being computed only once per timestep and not once per iteration. Do you agree on this or have another explanation? And if so, is there a way to update the stiffness tensor inside the `@Integrator` block?

Probably, yes. If you can't live with that (I would not), then we must make things a bit differently (the implementation of the consistent tangent operator becomes a more complex)

**Reply by kschnelll** [reply]
Hey Thomas, sorry for my late reply. You can find our code attached. 
I think the differentiation between yielding and compacting works fine for our needs, because in our case, the material is either compacting or yielding, usually not both... 
So the tangent operator is always computed? I thought it's only active when using the `@TangentOperator` keyword. 
The exponential adaption of the elastic modulus is pretty important for us, so I'm thankful for every tip you could give us. 

Thanks in advance, 
Klaus

[20221129_GitHub_helfer.txt](https://github.com/thelfer/tfel/files/10114753/20221129_GitHub_helfer.txt)
